lib/ex_typesense/document.ex

```diff
@@ -428,6 +452,97 @@ defmodule ExTypesense.Document do
+    HttpClient.request(conn, opts)
+  end
+
+  @doc """
+  Deletes documents in a collection by query.
+
+  > #### [Filter and batch size](https://typesense.org/docs/latest/api/documents.html#delete-by-query) {: .info}
+  >
+  > To delete all documents in a collection, you can use a filter that
+  > matches all documents in your collection. For eg, if you have an
+  > int32 field called popularity in your documents, you can use
+  > `filter_by: "popularity:>0"` to delete all documents. Or if you have a
+  > bool field called `in_stock` in your documents, you can use
+  > `filter_by: "in_stock:[true,false]"` to delete all documents.
+  >
+  > Use the `batch_size` to control the number of documents that should
+  > deleted at a time. A larger value will speed up deletions, but will
+  > impact performance of other operations running on the server.
+  >
+  > Filter parameters can be found here: https://typesense.org/docs/latest/api/search.html#filter-parameters
+
+  ## Examples
+      iex> query = %{
+      ...>   filter_by: "num_employees:>100",
+      ...>   batch_size: 100
+      ...> }
+      iex> ExTypesense.delete_documents_by_query(Employee, query)
+      {:ok, %{}}
+  """
+  @doc since: "0.5.0"
+  @spec delete_documents_by_query(
+          Connection.t(),
+          module() | String.t(),
+          %{
+            filter_by: String.t(),
+            batch_size: integer() | nil
+          }
+        ) ::
+          response()
+  def delete_documents_by_query(conn \\ Connection.new(), collection_name, query)
+
+  def delete_documents_by_query(conn, collection_name, %{filter_by: filter_by} = query)
+      when not is_nil(filter_by) and is_binary(filter_by) and is_atom(collection_name) do
+    collection_name = collection_name.__schema__(:source)
+    path = Path.join([@collections_path, collection_name, @documents_path])
+    HttpClient.request(conn, %{method: :delete, path: path, query: query})
+  end
+
+  def delete_documents_by_query(conn, collection_name, %{filter_by: filter_by} = query)
+      when not is_nil(filter_by) and is_binary(filter_by) and is_binary(collection_name) do
+    path = Path.join([@collections_path, collection_name, @documents_path])
+    HttpClient.request(conn, %{method: :delete, path: path, query: query})
   end
+
+  @doc """
+  Deletes all documents in a collection.
+
+  > #### On using this function {: .info}
+  > As of this writing (v0.5.0), there's no built-in way of deleting
+  > all documents via [Typesense docs](https://github.com/typesense/typesense/issues/1613#issuecomment-1994986258).
+  > This function uses `delete_by_query` under the hood.
+  """
+  @doc since: "0.5.0"
+  @spec delete_all_documents(Connection.t(), module() | String.t()) :: response()
+  def delete_all_documents(conn \\ Connection.new(), collection_name)
+
+  def delete_all_documents(conn, collection_name) when is_binary(collection_name) do
+    delete_documents_by_query(conn, collection_name, %{filter_by: "#{collection_name}_id:>=0"})
```

Comment:
**question:** In my case, the ID field of a document is called `id` and not `{collection_name}_id`. And I believe in most cases the collection name will not be prefixed to the `id` field name. So is this a conscious decision you made to prefix the collection name here?

Also, in the github comment thread you attached in the function doc, it says that `id>=0` will not work as the ID is a string field. I am sorry I missed this out while reviewing the changes. What do you think can be done here instead? @jaeyson

Reply:
I think a query that can work is `%{filter_by: "id:!=0"}`. This way it doesnt matter that the `id` is not an integer.

Reply:
@shijithkjayan good catch! I'll make changes to it.

Reply:
@shijithkjayan I forgot this part:

> To delete all documents in a collection, you can use a filter that matches all documents in your collection. For eg, if you have an int32 field called popularity in your documents, you can use filter_by=popularity:>0 to delete all documents. 

That's one of the reason why I used {collection_name}_id (type int32) instead if id (type string).

There's a thread discussion where a wildcard is not yet added as a feature:
https://threads.typesense.org/t/deleting-typesense-documents-with-specific-field-using-filter-by-/2J3912e

One problem here is if your PK in Ecto schema is a type of UUID (<-- I'll add this in my todo list)

Reply:
Even if its UUID, `%{filter_by: "id:!=0"}` will still work right?

Reply:
## `id:!=0`

@shijithkjayan I'm afraid not. Encountered an error by specifically doing that (`id:!=0`).

see this CI test: https://github.com/jaeyson/ex_typesense/actions/runs/10260778543/job/28387381407

`{:error, "Not equals filtering is not supported on the `id` field."}` <-- `!=` not supported

If you read this thread https://threads.typesense.org/2J3912e:

> Can i delete that documents using `filter_by: "user_id:*"`
> It's not possible to do a wildcard match on a field with filter by.

That's the reason why id or any fk field passed won't work at the moment when using wildcard for `filter_by`.

## uuid

> Even if its UUID

The reason why I brought up `uuid` is because currently I'm prefixing collection name (table name in Ecto) in `_id` and using it as default_sorting_field

Basing on their docs https://typesense.org/docs/26.0/api/collections.html#schema-parameters, it says in `default_sorting_field`:

> The name of an int32 / float field that determines the order in which the search results are ranked when a sort_by clause is not provided during searching.

So I'm not sure if a `default_sorting_field` with type `uuid` would work in this case.

Hope this answers your questions.

Reply:
I was actually able to clear all documents in my collection with `%{filter_by: "id:!=0"}` 🤔 

` ExTypesense.delete_documents_by_query(collection_name, %{filter_by: "id:!=0"}) ` --> This is what I used and it cleared everything with no errors

Reply:
@shijithkjayan what typesense version are you using?

This error `{:error, "Not equals filtering is not supported on the `id` field."}` is from using Typesense v0.25.0 on the test run result https://github.com/jaeyson/ex_typesense/actions/runs/10260778543/job/28387381407

The reason I don't want to continue to make on that change (`id:!=0`) is because it's not supported on v0.25.0.

I wanted this library to be backwards-compatible like atleast 1 or 2 versions below. I'll gladly change this based on your request once the version of Typesense bumps up, and not breaking on previous versions.